packages/docusaurus-theme-classic/src/theme/hooks/useTOCHighlight.ts

```diff
@@ -5,94 +5,115 @@
  * LICENSE file in the root directory of this source tree.
  */
 
-import {useEffect, useState} from 'react';
+import {Params} from '@theme/hooks/useTOCHighlight';
+import {useEffect, useRef} from 'react';
 
-function useTOCHighlight(
-  linkClassName: string,
-  linkActiveClassName: string,
-  topOffset: number,
-): void {
-  const [lastActiveLink, setLastActiveLink] = useState<
-    HTMLAnchorElement | undefined
-  >(undefined);
+// If the anchor has no height and is just a "marker" in the dom; we'll use the parent (normally the link text) rect boundaries instead
+function getVisibleBoundingClientRect(element: HTMLElement): DOMRect {
+  const rect = element.getBoundingClientRect();
+  const hasNoHeight = rect.top === rect.bottom;
+  if (hasNoHeight) {
+    return getVisibleBoundingClientRect(element.parentNode as HTMLElement);
+  }
+  return rect;
+}
+
+// Considering we divide viewport into 2 zones of each 50vh
+// This returns true if an element is in the first zone (ie, appear in viewport, near the top)
+function isInViewportTopHalf(boundingRect: DOMRect) {
+  return boundingRect.top > 0 && boundingRect.bottom < window.innerHeight / 2;
+}
+
+function getAnchors() {
+  // For toc highlighting, we only consider h2/h3 anchors
```

Comment:
In the future, we'll make headings levels displayed in TOC customizable, so maybe now we can avoid hard-coded h2/h3 only?

Reply:
too late :D 

Agree that this should be configurable, but let's see how to add this in another PR like https://github.com/facebook/docusaurus/pull/4310

Reply:
I think anchors created in TOC should have an extra class like `anchor__toc` or something, because otherwise we'd still get weird behaviors if user render headings with JSX/React and they don't appear in the TOC